src/modules/fw_pos_control/FixedwingPositionControl.cpp

```diff
@@ -2709,6 +2718,32 @@ FixedwingPositionControl::tecs_update_pitch_throttle(const float control_interva
 		     hgt_rate_sp);
 
 	tecs_status_publish(alt_sp, airspeed_sp, airspeed_rate_estimate, throttle_trim_adjusted);
+
+	if (_tecs_is_running && !_vehicle_status.in_transition_mode
+	    && (_vehicle_status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_FIXED_WING)) {
+		// Check level flight: the height rate setpoint is not set or set to 0 and we are close to the target altitude and target altitude is not moving
+		if ((!PX4_ISFINITE(hgt_rate_sp) || fabsf(hgt_rate_sp) < MAX_ALT_REF_RATE_FOR_LEVEL_FLIGHT) &&
+		    fabsf(_current_altitude - alt_sp) < _param_nav_fw_alt_rad.get() &&
+		    fabsf((alt_sp - _last_tecs_alt_sp) / control_interval) < MAX_ALT_REF_RATE_FOR_LEVEL_FLIGHT) {
```

Comment:
Ideally it would not be couple with the navigator acceptance logic, but I wouldn't have a good idea how to do it else.
You on purpose don't want to rely on `tecs_status.altitude_reference`?

Reply:
That is the internally filtered altitude reference i chose the setpoint since when you command a new one, it immediately switches to the climb flight phase and not after the filtered one ramped up. So the flight phase is more coupled to the commands not not the internal filter. oth would be viable though.